src/lib.rs

```diff
@@ -256,71 +261,89 @@ impl SubnetActor for Actor {
-    //     let mut st = State::load();
-    //     let caller = Address::new_id(sdk::message::caller());
-    //     // check type of caller
-    //     let code_cid = get_actor_code_cid(&caller).unwrap_or(Cid::default());
-    //     if sdk::actor::get_builtin_actor_type(&code_cid) != Some(Type::Account) {
-    //         abort!(USR_FORBIDDEN, "caller not account actor type");
-    //     }
-    //
-    //     let ch_cid = checkpoint.cid();
-    //     // verify checkpoint
-    //     st.verify_checkpoint(&checkpoint)?;
-    //
-    //     // get votes for committed checkpoint
-    //     let mut votes_map = make_map_with_root::<_, Votes>(&st.window_checks, &Blockstore)
-    //         .map_err(|e| anyhow!("failed to load checkpoints: {}", e))?;
-    //     let mut found = false;
-    //     let mut votes = match get_votes(&votes_map, &ch_cid)? {
-    //         Some(v) => {
-    //             found = true;
-    //             v.clone()
-    //         }
-    //         None => Votes {
-    //             validators: Vec::new(),
-    //         },
-    //     };
-    //
-    //     if votes.validators.iter().any(|x| x == &caller) {
-    //         return Err(anyhow!("miner has already voted the checkpoint"));
-    //     }
-    //
-    //     // add miner vote
-    //     votes.validators.push(caller);
-    //
-    //     // if has majority
-    //     if st.has_majority_vote(&votes)? {
-    //         // commit checkpoint
-    //         st.flush_checkpoint::<&Blockstore>(&checkpoint)?;
-    //         // propagate to sca
-    //         st.send(
-    //             &Address::new_id(sca_actor_addr),
-    //             Method::CommitChildCheckpoint as u64,
-    //             RawBytes::serialize(checkpoint)?,
-    //             0.into(),
-    //         )?;
-    //         // remove votes used for commitment
-    //         if found {
-    //             votes_map.delete(&ch_cid.to_bytes())?;
-    //         }
-    //     } else {
-    //         // if no majority store vote and return
-    //         votes_map.set(ch_cid.to_bytes().into(), votes)?;
-    //     }
-    //
-    //     // flush votes
-    //     st.window_checks = votes_map.flush()?;
-    //
-    //     st.save();
-    //     Ok(None)
-    // }
+    /// SubmitCheckpoint accepts signed checkpoint votes for miners.
+    ///
+    /// This functions verifies that the checkpoint is valid before
+    /// propagating it for commitment to the IPC gateway. It expects at least
+    /// votes from 2/3 of miners with collateral.
+    fn submit_checkpoint<BS, RT>(
+        rt: &mut RT,
+        ch: Checkpoint,
+    ) -> Result<Option<RawBytes>, ActorError>
+    where
+        BS: Blockstore,
+        RT: Runtime<BS>,
+    {
+        let state: State = rt.state()?;
+        let caller = rt.message().caller();
+
+        if !state.is_validator(&caller) {
+            return Err(actor_error!(illegal_state, "not validator"));
+        }
+
+        state
```

Comment:
What about addressing the issue you raised [here](https://github.com/consensus-shipyard/ipc-subnet-actor/issues/10) already in this PR?

Reply:
I'll open another separate PR for this one, is it ok?

Reply:
Sure, no problem. Let's land this one then and have that as a follow-up.